onnxruntime/core/providers/rocm/math/binary_elementwise_ops.h

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/providers/rocm/rocm_common.h"
+#include "core/providers/rocm/shared_inc/fast_divmod.h"
+#include "core/providers/cpu/tensor/utils.h"
+
+namespace onnxruntime {
+namespace rocm {
+
+struct BinaryElementwisePreparation {
+  const Tensor* lhs_tensor = nullptr;
+  const Tensor* rhs_tensor = nullptr;
+  Tensor* output_tensor = nullptr;
+  int32_t output_rank_or_simple_broadcast = 0; // for no_broadcast|left_scalar|right_scalar cases, output_rank uses SimpleBroadcast enums
+
+  RocmKernel::HipAsyncBuffer<int64_t> lhs_padded_strides;  // for lhs shape == output shape, this is nullptr
```

Comment:
Perhaps it would be worth adding a comment about why TArray was avoided? In other words the perf issue about passing struct by value and that it's expected fixed in the future.

Reply:
agree.